developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+suspend the interface, which then breaks following activation attempts. The
+current VirtIO implementation contains a fix for that, but that fix might turn
+out to be incorrect. The difficulty in establishing a correct fix is, that so
+far it hasn't been possible to find a test case for
+snd_soc_runtime_deactivate().
+
+3.3. Deactivation
+-----------------
+
+When the guest isn't actively using audio, we want to allow the host to
+runtime suspend the DSP. That usually means switching the DSP off. Next time the
+guest wants to activate audio it will have to perform initialisation again,
+including sending its topology to the host. This happens automatically as long
+as the guest runtime suspends too or performs the initialisation cycle in some
+other way, by unloading and reloading drivers, or by rebooting.
+
+With automatic runtime suspending on the host and the guest, it can happen,
+that one of them got suspended, whereas the other one is still running. A
+problematic case would be, if the host got suspended, while the guest was still
+active, because then, when the host resumes, it will also bring up the DSP and
+only send the host topology part to it. The guest however wasn't suspended,
+therefore it won't send its topology to the host. To avoid this the automatic
+runtime suspend timeout on the guest is set to a half of the host's timeout.
+
+The opposite case also requires special handling: when the host is active and
+the guest got suspended and is sending its topology to the host. If the host
+then forwards that topology to the DSP as usual, the DSP firmware will detect
+component duplication and enter an erroneous state. These cases also cannot be
+reliably avoided, because it's always possible, that while the guest's audio is
+idle, the host's audio is still active, e.g. being in use locally. To handle
+such cases correctly, we implement a reset count. The host counts DSP resets,
+and when the guest is resuming, it requests a reset status, using a newly added
+SOF_IPC_PM_VFE_RESET_STATUS IPC message. If the guest identifies, that the host
+hasn't been reset since the guest's last suspend, it skips sending the parsed
+topology to it.
+
+Guest reboot also requires special handling to deinitialise virtual queues. At
+the moment there is no standard way to get notified about the guest reboot event
+on the host. Various vhost drivers implement this in their own context-specific
+ways. E.g. the vhost networking driver uses a link status update for that
+purpose. However the vsock vhost driver comes very close to our needs. It
+doesn't have a suitable context notification, therefore it implements a
+dedicated QEMU / misc-device ioctl VHOST_VSOCK_SET_RUNNING. We reuse the ioctl
+definition, but change its name to VHOST_SET_RUNNING. This way no existing apps
+have to be modified. Then we use this ioctl() in QEMU SOF code to inform the SOF
+vhost driver about a guest reboot. This is then used to reset VirtQueue status
+on the host.
```

Comment:
We also need to teardown and guest topology on the DSP.

Reply:
@lgirdwood is there a way to do that apart from resetting the DSP? Otherwise - is there any harm in it keeping the topology until the next reset, whenever that happens?

Reply:
yep, there is IPC to teardown pipelines. Your pipeline will be in guest component IPC ID range